### PR TITLE
ignore `llvm::Lld` if lld is not enabled

### PR DESCRIPTION
**Pull request by onur-ozkan**
People are having trouble ([ref. zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/MSVC.20Runtime.20mismatch.20when.20building.20LLD)) when they don't want to build `lld` for their custom distribution tarballs even with `lld = false` in their config.toml. This is because it is not controlled by `lld_enabled` flag. This change ensures that `llvm:Lld` is controlled by lld configuration.

Additionally, `lld = true` is set by default for dist profile, because we have been building it all along and this maintains that behavior.

try-job: x86_64-mingw

### CODE REVIEW

**Comment by onur-ozkan** [comment]
> This PR modifies `src/bootstrap/defaults`.
> 
> If appropriate, please update `CONFIG_CHANGE_HISTORY` in `src/bootstrap/src/utils/change_tracker.rs`.

`src/bootstrap/defaults/config.dist.toml` was updated to avoid a breaking change, so there is nothing to add to the change tracker.

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile [hunk 1]
```diff
@@ -51,6 +51,7 @@ ENV RUST_CONFIGURE_ARGS \
       --build=x86_64-unknown-linux-gnu \
       --llvm-root=/usr/lib/llvm-18 \
       --enable-llvm-link-shared \
+      --set rust.lld=false \
```
Comment (Kobzol) [inline src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile:54]: Hmm, if an external LLVM is used, it shouldn't even be possible to build LLD, IMO. This goes back to the discussion that perhaps the `lld` config should be under `llvm` and it should just mean "build LLD" and nothing more :) Anyway, that's just me thinking out loud.

Comment (onur-ozkan) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile:54]: > This goes back to the discussion that perhaps the lld config should be under llvm and it should just mean "build LLD" and nothing more :)

I think the same way, but the scope of this PR is kind a different (utilizing lld_enabled flag for dist tarball).

Comment (onur-ozkan) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile]: What do you think about https://github.com/rust-lang/rust/pull/126701/commits/7aec1b01bbb9d06956819f34371cf071858d2f31 ?

Comment (Kobzol) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-18/Dockerfile]: Oops, I completely missed this comment, sorry. Yeah, that makes sense. If we use an external LLVM, we shouldn't build LLD (that's also why LLD should be in the `llvm` section, IMO :) ).

**Comment by Kobzol** [comment]
This is a good cleanup, it should be possible to turn packaging of `lld` off. Let's do a try build just to check that the `lld` component stays there.

@bors try

**Comment by Kobzol** [comment]
Good, `lld` seems to be there!

One more thing, maybe we should add an entry to config tracker? Maybe some people who do their own `dist` would be interested in the changed default (they might have overridden `lld` before or they don't use the `dist` default profile).

File: src/bootstrap/src/utils/change_tracker.rs [hunk 2]
```diff
@@ -195,4 +195,9 @@ pub const CONFIG_CHANGE_HISTORY: &[ChangeInfo] = &[
         severity: ChangeSeverity::Warning,
         summary: "Removed `dist.missing-tools` configuration as it was deprecated long time ago.",
     },
+    ChangeInfo {
+        change_id: 126701,
+        severity: ChangeSeverity::Warning,
+        summary: "`llvm.lld` is enabled by default for the dist profile. If set to false, `lld` will not be included in the build.",
```
Comment (Kobzol) [inline src/bootstrap/src/utils/change_tracker.rs:201]:
```suggestion
        summary: "`llvm.lld` is enabled by default for the dist profile. If set to false, `lld` will not be included in the dist build.",
```
Just to make it clear that this affects primarily `x dist`?

**Comment by tgross35** [comment]
Looks like this might have failed the rollup through the test rust-lld-by-default https://github.com/rust-lang/rust/pull/127067#issuecomment-2196101499

**Comment by onur-ozkan** [comment]
> Looks like this might have failed the rollup through the test rust-lld-by-default [#127067 (comment)](https://github.com/rust-lang/rust/pull/127067#issuecomment-2196101499)

`lld` was disabled because the runner uses `stable` as a release channel: https://github.com/rust-lang/rust/blob/42add88d2275b95c98e512ab680436ede691e853/src/ci/github-actions/jobs.yml#L224-L226

I [improved the logic](https://github.com/rust-lang/rust/pull/126701/commits/371221174ab915c9f61117833af9c762be95a407) where we enable it in bootstrap.

cc @lqd

**Review by Kobzol — COMMENTED** [review]
I *think* that this is what has happened here:
- `dist` profile now enables `lld`
- this enables the `rust-lld-by-default` test, because before it was not enabled, since it uses `//@ needs-rust-lld`
- the test failed on stable (which it should)

Note: the `dist` step has packaged `lld` before, even if `lld` wasn't enabled in `config.toml`. So this PR has changed the behavior of tests, since now we respect `lld` in `config.toml` when packaging, but that means that the lld test probably needs an update.

**Comment by Kobzol** [comment]
Good! Now let's check if it has helped, just to be sure.

@bors try

**Comment by Kobzol** [comment]
Okay, looks good. Let's try again.

@bors r+ rollup=never

**Comment by Kobzol** [comment]
Yet another new test enabled by this PR. Fun fact, this test was ignored on all builders before this PR :upside_down_face:

**Comment by Kobzol** [comment]
It was ignored because of `//@ needs-rust-lld`. Probably it needs `//@ only-msvc` instead of `//@ only-windows` then?

**Comment by lqd** [comment]
Ah so likely yet another case of rust-lld was only enabled on some dist builders and no test builder.

Probably something like that directive yes, or maybe even more precise à la x86_64-pc-windows-msvc, as IIUC there are a bunch of other msvc targets that may not have rust-lld support (or where it's distributed though I guess needs-rust-lld will take care of that). But I know nothing about windows and this may need the eyes of e.g. Chris Denton. 

That is, if the test actually still passes and hasn't regressed since it was written :/ Maybe you can try it on x64 msvc and it it fails ping Chris the expert?

**Comment by Kobzol** [comment]
The test has succeeded on `x86_64-msvc` in the previous merge attempt. So probably we just need to modify the compiletest flag.

**Comment by lqd** [comment]
great then `//@ only-x86_64-pc-windows-msvc` and we're good

**Comment by onur-ozkan** [comment]
Testing https://github.com/rust-lang/rust/pull/126701/commits/17b843bc2cc2db2bf86f2c93ec3bade5f358ba57

@bors try

**Comment by onur-ozkan** [comment]
> @bors r+ rollup=never

oops, didn't see that

**Pull request merged**